### ISSUE TITLE
Proposal (& implementation): LINQ APIs to enable C# 8.0 index and range for IEnumerable<T>

### ISSUE THREAD

**Issue by Dixin** [dotnet/runtime#28776]
C# 8.0 introduces index and range features for array and [countable types](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/ranges#adding-index-and-range-support-to-existing-library-types). It is natural and convenient to **generally support index and range for all IEnumerable<T> types and LINQ queries**. I searched the [API review notes](https://github.com/dotnet/apireviews), didn't find such APIs, so propose them here.

```cs
var element = source1.ElementAt(index: ^5);
var elements = source2.Take(range: 10..^10).Select().Where();
```

# Problem

Index/range are **language level** features. Currently, they

- work with array (compiled to `RuntimeHelpers.GetSubArray`, etc.)
- work with [countable type](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/ranges#adding-index-and-range-support-to-existing-library-types)
- do not work with "uncountable" `IEnumerable<T>` sequence, or LINQ query.

# Rationale and usage

The goals of these LINQ APIs are:

- Use index to locate an element in sequence.
- Use a range to slice a sequence. The usage should be consistent with array, but with deferred execution.

This enables index and range **language features** to work with any type that implements `IEnumerable<T>`, and LINQ queries.

LINQ already has `ElementAt(int)` and `ElementAtOrDefault(int)` operators. It would be natural to have a `System.Index` overload as `ElementAt(Index)` and `ElementAtOrDefault(Index)`, as well as `Slice(Range)`, so that LINQ can seamlessly work with C# 8.0:

```cs
Index index = ...;
var element1 = source1.ElementAt(index);
var element2 = source2.ElementAtOrDefault(^5);
Range range = ...;
var slice1 = source3.Take(range);
var slice2 = source4.Take(2..^2);
var slice3 = source5.Take(^10..);
```

With these APIs, the C# `countable[Index]` and `countable[Range]` syntax are enabled for sequences & LINQ queries as `enumerable.ElementAt(Index)` and `enumerable.Slice(Range)`.

# Proposed APIs

For LINQ to Objects:

```cs
namespace System.Linq
{
    public static partial class Enumerable
    {
        public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, Index index);

        public static TSource ElementAtOrDefault<TSource>(this IEnumerable<TSource> source, Index index);

        public static IEnumerable<TSource> Take<TSource>(this IEnumerable<TSource> source, Range range);
    }
}
```

For remote LINQ:

```cs
namespace System.Linq
{
    public static partial class Queryable
    {
        public static TSource ElementAt<TSource>(this IQueryable<TSource> source, Index index);

        public static TSource ElementAtOrDefault<TSource>(this IQueryable<TSource> source, Index index);

        public static IQueryable<TSource> Take<TSource>(this IQueryable<TSource> source, Range range);
    }
}
```

**API review feedback**: @eiriktsarpalis:
> We'll be implementing the Queryable equivalents but not the PLINQ ones.

# Implementation details (and pull request)

The [API review process](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md) says PR should not be submitted before the API proposal is approved. So currently I implemented these APIs separately [Dixin/Linq.IndexRange](https://github.com/Dixin/Linq.IndexRange):

- `Enumerable`:
  - [`ElementAt(Index)` and `ElementAtOrDefault(Index)`](https://github.com/Dixin/System.Linq.IndexRange/blob/master/System.Linq.IndexRange/ElementAt.cs)
  - [`Slice(Range)`](https://github.com/Dixin/System.Linq.IndexRange/blob/master/System.Linq.IndexRange/Slice.cs)
- `Queryable`:
  - [`ElementAt(Index)`, `ElementAtOrDefault(Index)`, `Slice(Range)`](https://github.com/Dixin/System.Linq.IndexRange/blob/master/System.Linq.IndexRange/QueryableExtensions.cs).

Please see the [unit tests](https://github.com/Dixin/System.Linq.IndexRange/tree/master/System.Linq.IndexRange.Tests) about how they work.

These proposed APIs can be used by adding NuGet package `Linq.IndexRange`:

```cmd
dotnet add package Linq.IndexRange
```

If this proposal is doable, I can submit a PR quickly.

# Open questions

## `Slice(Range)` vs `ElementsIn(Range)` as name

Should `Slice` be called `ElementsIn(Range)`? Currently I implemented both.

- `ElementsIn(Range)` keeps the naming consistency with original `ElementAt(Index)`. Might it be natural for existing LINQ users?
- `Slice` is consistent with the [countable types](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/ranges#adding-index-and-range-support-to-existing-library-types), which requires a `Slice` method to support range. I prefer `Slice` for this reason.

**API review feedback**: Rename to `Take(Range)`.
@eiriktsarpalis:
> I'm increasingly warming up to exposing the `Range` method as a `Take` overload, it really simplifies applications that would typically require chaining `Skip` and `Take` calls and the confusion that ensues with order of chaining and indices:
> 
> * `source.Take(..3)` instead of `source.Take(3)`
> * `source.Take(3..)` instead of `source.Skip(3)`
> * `source.Take(2..7)` instead of `source.Take(7).Skip(2)`
> * `source.Take(^3..)` instead of `source.TakeLast(3)`
> * `source.Take(..^3)` instead of `source.SkipLast(3)`
> * `source.Take(^7..^3)` instead of `source.TakeLast(7).SkipLast(3)`. This is also more efficient since elements only need to be queued once instead of twice.
> 
> Using the name `Take` also removes any ambiguity in terms of whether out-of-bounds ranges should have `Slice` semantics or `Take` semantics. Clearly, it should be the latter. The semantics of the [implementation as provided by @Dixin](https://github.com/Dixin/System.Linq.IndexRange/blob/c66ba6090cd3bc1dfee8efe1cf359ffadaf617f7/System.Linq.IndexRange/Slice.cs) are just right in that regard.

## Out of bound handling

If Index is out of the boundaries, for `ElementAt(Index)`, there are 2 options:

- **Array behavior**: throw `IndexOutOfRangeException`
- **LINQ behavior**: throw `ArgumentOutOfRangeException`. My current implementation goes this way, to keep `ElementAt(Index)` consistent with orginal `ElementAt(int)`.

If Range goes off the boundaries of source sequence, for `Slice(Range)`, there are 2 options:

- **Array behavior**: Follow the behavior of `array[Range]`, throw `ArgumentOutOfRangeException`. I implemented `ElementsIn(Range)` following this way. See [unit tests of `ElementsIn`](https://github.com/Dixin/System.Linq.IndexRange/blob/master/System.Linq.IndexRange.Tests/ElementsInTests.cs).
- **LINQ behavior**: Follow the behavior of current partitioning LINQ operators like `Skip`/`Take`/`SkipLast`/`TakeLast`, do not throw any exception. I implemented `Slice` following this way. See [unit test of `Slice`](https://github.com/Dixin/System.Linq.IndexRange/blob/master/System.Linq.IndexRange.Tests/SliceTests.cs).

**API review feedback**: @eiriktsarpalis:
> We use LINQ behaviour for out-of-bounds ranges.

> Using the name `Take` also removes any ambiguity in terms of whether out-of-bounds ranges should have `Slice` semantics or `Take` semantics. Clearly, it should be the latter.

## `ElementAt(Index)` and `Queryable`

As @bartdesmet mentioned in the comments, LINQ providers may have issues when they see `ElementAt` having an `Index` argument, etc. Should we have a new name for the operator instead of overload? For example, `At(Index)` or `Index(Index)`?

# Feedback

My original proposal includes other methods like `Enumerable.Range(Range)` and `AsEnumerable(this Range)`. After discussion with @eiriktsarpalis, we are keeping only `ElementAt(Index)` and `Slice(Range)`. The original proposal can be found in the history or [this link](https://github.com/Dixin/System.Linq.IndexRange).

I had an email discussion with @MadsTorgersen, merging his feedback here:
> Thanks for looping me in, and for the great proposal!

> A couple of thoughts:
> - I like the proposal! I have no concerns about adding this functionality to IEnumerable<T>. I think that the specific method names should be considered in the larger context of standardizing them for pattern-based compilation, but that’s a detail.
> - I agree with most of Bart’s comments on the proposal, including the open questions he points out.

# API review feedback

@eiriktsarpalis:
> - We're renaming Slice to Take.
> - We use LINQ behaviour for out-of-bounds ranges.
> - We'll be implementing the Queryable equivalents but not the PLINQ ones.

**Comment by eiriktsarpalis** [comment]
> For Range(Range) and AsEnumerable(Range), the question is: what does range's start index and end index mean, when the index is from the end? For example, 10..20 can be easily converted to a sequence of 10, 11,12, ... 19, but how about ^20...^10?
>
>The easiest way is to disallow, and throw exception.
>
>My current implementation attempts to make it flexible. Regarding Index's Value can be from 0 to int.MaxValue, I assume a virtual "full range" 0..2147483648, and any Range instance is a slice of that "full range". So:
>
>* Ranges .. and 0.. and ..^0 and 0..^0 are converted to "full sequence" 0, 1, .. 2147483647
>* Range 100..^47 is converted to sequence 100, 101, .. 2147483600
>* Range ^48..^40 is converted to sequence 2147483600, 2147483601 .. 2147483607
>* Range 10..10 is converted to empty sequence

My concern with this approach is that it seems to assign arbitrary meaning to "from end" indices simply to reconcile them with IEnumerable. While throwing is probaby the better approach, it can still be violating user expectations, particularly if the type implementing IEnumerable is a list:
```csharp
var l = new List<int>() { 1, 2, 3 };

l[^1]; // 3
l.ElementAt(^1); // throws ArgumentOutOfRangeException
```
It could be argued that a type test could be used and adapt meaning if the underlying type is `IList`, however this would imply completely changing semantics depending on the subtype. This approach has LSP violation written all over it.

As such, I would recommend against accepting this proposal. We should simply acknowledge that indices and ranges are intended for list-like types, and IEnumerable is much more general-purpose than that.

**Comment by Dixin** [comment]
@eiriktsarpalis, thank you for your review and feedback. Actually in my proposed implementation, the meaning of "from end" is consistent with existing APIs.

For the proposed static method `Enumerable.Range(Range)` API:
- `Enumerable.Range(2..7)`  should make sense.
- `Enumerbale.Range(^10..^0)` may be violate user expectation, as you pointed out.

I agree this particular method should only allow "from start", and disallow "from end".

For the other method `Enumerable.ElementAt(Index)` that you mentioned, in my [proposed implementation](url), it allows "from start" and "from end" index, and it works exactly the same way as the countable types.
```
var countable = new List<int>() { 1, 2, 3 };
var enumerable = countable.Hide();

countable[^1] // 3
enumerable.ElementAt(^1) // 3

countable[^10] // ArgumentOutOfRangeException
enumerable.ElementAt(^10) // ArgumentOutOfRangeException
```

You can also take a look at the [unit tests](https://github.com/Dixin/System.Linq.IndexRange/tree/master/System.Linq.IndexRange.Tests) of the proposed implementation.

You mentioned
> IEnumerable is much more general-purpose than that

That's why I proposed these APIs. They would enable Index and Range for so many types in .NET and for LINQ queries. If they keep consistency with existing APIs, they can provide good convenience.

Earlier I had an email discussion with @MadsTorgersen, quoting his feedback:
> Thanks for looping me in, and for the great proposal! 

> A couple of thoughts:
> - I like the proposal! I have no concerns about adding this functionality to IEnumerable<T>. I think that the specific method names should be considered in the larger context of standardizing them for pattern-based compilation, but that’s a detail.
> - I agree with most of Bart’s comments on the proposal, including the open questions he points out.

Thank you @eiriktsarpalis so much again for your time. Please let me know what you think.

**Comment by eiriktsarpalis** [comment]
Thanks for the clarification @Dixin, I mistook your original comment as specifying the "from end" behaviour for all proposed methods and not just `Enumerable.Range`.

The `ElementAt` implementation that you linked to seems reasonable. Only downside being that "from end" index support implies _O(n)_ space, however it's probably not that important for the basic applications the method is meant to be used in (though the potential for misuse will probably not go away!)

Going back to `Enumerable.Range`, are we comfortable with the fact that `Enumerable.Range(5 .. 10)` evaluates to something completely different than `Enumerable.Range(5, 10)`? While it may be obvious to people who read the docs or are familiar with .NET nuance and history, for casual users it might not be so. Perhaps using a different method name is preferable here? (for instance the proposed `AsEnumerable`).

I tend to favour `Slice` over `ElementsIn` and would probably prefer non-throwing behaviour for out-of-bounds ranges.

Also, @Dixin, it looks most of the `Linq.IndexRange` links in your OP are dead. For the purposes of assisting future review, would it be possible to update to permalinks? Thanks.

**Comment by eiriktsarpalis** [comment]
> Supporting "from end" is a different matter though. On the one hand, given the int return type, one could "suspect" a range 0..int.MaxValue, but a more formally inclined person could argue about the signed nature of int (int.MinValue..int.MaxValue?).

Note that ranges only support non-negative numbers, since they are ranges of indices rather than a general-purpose range construct.

**Comment by Dixin** [comment]
@eiriktsarpalis, thank you for your reply.

> The ElementAt implementation that you linked to seems reasonable. Only downside being that "from end" index support implies O(n) space, however it's probably not that important for the basic applications the method is meant to be used in

That is by design. 
- If index is from start `ElementAt(Index)` is the same as the original `ElementAt(int)`. 
- If index is from end, it does take O(n) space, because we do not know the count of the enumerable. 
  - `enumerable.ElementAt(^5)` is equivalent to `enumerable.TakeLast(5).ElementAt(0)`. Regarding consistency & performance, I am following the approach in `TakeLast` operator implementation.
  - To make it O(1) space, we have to scan the enumerable twice, that not how enumerable supposed to work.

> Going back to Enumerable.Range, are we comfortable with the fact that Enumerable.Range(5 .. 10) evaluates to something completely different than Enumerable.Range(5, 10)? While it may be obvious to people who read the docs or are familiar with .NET nuance and history, for casual users it might not be so. Perhaps using a different method name is preferable here? (for instance the proposed AsEnumerable).

I totally realize the new `Enumerable.Range(5 .. 10)` is totally different from the old `Enumerable.Range(5, 10)`. On the one hand I am also thinking, because C# officially introduces the Range concept, would `Enumerable.Range(Range)` be more natural for modern C# users?

**Comment by Clockwork-Muse** [comment]
> * If index is from end, it does take O(n) space, because we do not know the count of the enumerable.
>   
>   * `enumerable.ElementAt(^5)` is equivalent to `enumerable.TakeLast(5).ElementAt(0)`. Regarding consistency & performance, I am following the approach in `TakeLast` operator implementation.

Technically this equivalent proposal only requires O(5) space, at most.  There's no reason a from-end `ElementAt` wouldn't be able to take the same shortcuts.

**Comment by Dixin** [comment]
> Also, @Dixin, it looks most of the Linq.IndexRange links in your OP are dead. For the purposes of assisting future review, would it be possible to update to permalinks?

@eiriktsarpalis I updated the proposal contents including all links. Thank you.

**Comment by Dixin** [comment]
> Technically this equivalent proposal only requires O(5) space, at most. There's no reason a from-end `ElementAt` wouldn't be able to take the same shortcuts.

@Clockwork-Muse Yes, the proposed implementation only requires O(5) space, at most. Here is the [link to the source](https://github.com/Dixin/System.Linq.IndexRange/blob/master/System.Linq.IndexRange/ElementAt.cs).

**Comment by eiriktsarpalis** [comment]
> > Technically this equivalent proposal only requires O(5) space, at most. There's no reason a from-end ElementAt wouldn't be able to take the same shortcuts.
>
>@Clockwork-Muse Yes, the proposed implementation only requires O(5) space, at most. Here is the link to the source.

It's _Θ(k)_ for `ElementAt(^k)` calls, so _O(n)_ space by virtue of the fact that `source.ElementAt(^source.Count())` is valid.

**Comment by Dixin** [comment]
> I'm not too sure about that.

I am not sure either, and I don't have a preference here, just trying to put everything on the plate for discussion. @eiriktsarpalis please let me know if you want me to remove it from the proposal or make other changes.

**Comment by eiriktsarpalis** [comment]
At this point I'm fairly convinced that we should hold of from adding the `Enumerable.Range` and `AsEnumerable` methods. I think adding the `Slice` and `ElementAt*` methods is probably fine despite the performance concerns (the existing `ElementAt` method isn't too great in that regard either) since they are consuming index types the way they were intended.

**Comment by Dixin** [comment]
> I think adding the `Slice` and `ElementAt*` methods is probably fine despite the performance concerns (the existing `ElementAt` method isn't too great in that regard either) since they are consuming index types the way they were intended.

@eiriktsarpalis Thank you for your insight. Can I submit a PR for `Slice` and `ElementAt*`? Or should I wait for further approval?

**Comment by eiriktsarpalis** [comment]
It would need to go through [API review](https://github.com/dotnet/runtime/blob/master/docs/project/api-review-process.md) before we can consider an implementation. Would it be possible to update the proposal to only include `Slice` and `ElementAt*`?

**Comment by Dixin** [comment]
> Before I mark this as ready for review, would it be possible to update the proposal to only include `Slice` and `ElementAt*`?

@eiriktsarpalis It's updated. Please let me know if more modification is needed. Thank you for taking this to further review. This is very exciting!

**Comment by eiriktsarpalis** [comment]
Thanks, main open question for me is whether we would also want to add the queryable APIs. @bartdesmet is right to point out that it may break query providers if they rely on the fact that there are no overloads for `ElementAt`, but I'm not sure if there is precedent here cc @stephentoub

**Comment by terrajobst** [comment]
* We're unsure about `Slice`, for two reasons:
    - We generally want `Slice` to return the same type it's invoked on (hard for interfaces)
    - It's unclear whether the API should have `Take` semantics (where bounds are min/max) or `Slice` where the enumerable must have that exact range
* We're also unsure about `Queryable`, some of us believe we evolved `Queryable` already, but we should make sure that's the case an inline with our breaking change policy on Linq providers
* Should we evolve PLinq as well?

```C#
namespace System.Linq
{
    public static partial class Enumerable
    {
        public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, Index index);
        public static TSource? ElementAtOrDefault<TSource>(this IEnumerable<TSource> source, Index index);
        public static IEnumerable<TSource> Slice<TSource>(this IEnumerable<TSource> source, Range range);
    }

    public static partial class Queryable
    {
        public static TSource ElementAt<TSource>(this IQueryable<TSource> source, Index index);
        public static TSource? ElementAtOrDefault<TSource>(this IQueryable<TSource> source, Index index);
        public static IQueryable<TSource> Slice<TSource>(this IQueryable<TSource> source, Range range);
    }
}
```

**Comment by jogibear9988** [comment]
EF Core reflection needs to be fixed when Queryable ElementAt is changed: https://github.com/dotnet/efcore/blob/5d537965c5a547c3d6e42b171d7f8a9415554093/src/EFCore/Query/QueryableMethods.cs#L477

**Comment by jogibear9988** [comment]
In linq2db it seems to be ok https://github.com/linq2db/linq2db/blob/37d0f108214c2fc0e6f1991c9d08627c605a9241/Source/LinqToDB/Reflection/Methods.cs#L41

**Comment by GrabYourPitchforks** [comment]
For the queryable part, what should happen when we have the following?

```cs
IQueryable<T> q = GetQueryable();
IQueryable<T> sliced = q.Slice(^10..30);
```

How would this get rewritten in terms of existing LINQ methods? I'm imagining some complex "do a _Take_ and a _TakeLast_ and figure out how to intersect them" magic, but doing all of this in an index-based (not value-based) way seems like it'll be very contorted without introducing new logic within the LINQ providers themselves. A failure of imagination on my part, perhaps?

**Comment by jogibear9988** [comment]
There is also

ElementAtAsync
ElementAtOrDefaultAsync

**Comment by eiriktsarpalis** [comment]
@GrabYourPitchforks presumably individual query providers could choose to reject "from end" indices and ranges to keep things sane.

**Comment by bartdesmet** [comment]
Query providers can definitely just reject the new methods at runtime during query translation, or only choose to support a subset of functionality. They can also resort to translation to corresponding Enumerable methods to run residual computation locally if that's not deemed to expensive. (This is commonly done for final protections but sometimes for other query operators as well.)

This said, Queryable overloads would have limited mileage without proper language support for from-end and range expressions. That is, the following would work:

xs.ElementAt(^1)

assuming xs is an IQueryable<T>, but only if it doesn't occur in an expression tree context, because the C# compiler does not support using ^ or .. in an expression tree (nor do the libraries support FromEnd or Range nodes in expression trees). For example:

from x in xs.Slice(1..2)
from y in ys.Slice(3..4)
...

would fail with an error highlighting 3..4 (but not 1..2) stating expression trees don't support range expressions. The reason why may not be entirely obvious to the author, but becomes clearer when expanding the query expression to

xs.Slice(1..2).SelectMany(x => ys.Slice(3..4))

Now it's clear that 3..4 occurs in the body of a lambda being converted to an expression tree, so it doesn’t work there.

Obviously it'd be great if C# would end up supporting newer language constructs in expression trees. It's a sore spot at the moment and the gap only widens. A possible implementation (with a Roslyn fork) is showcased at https://github.com/bartdesmet/ExpressionFutures/tree/master/CSharpExpressions. I've recently added support from ^ and .. in there as well, but it remains a "chasing taillights" hobby project (though with a ton of tests to ensure semantic equivalence between compiled C# code and runtime compilation of resulting expression trees).

But I digress... Even with the lack of language support, one could then resort to writing a new object creation expression with a Range constructor (or use factory methods), which LINQ providers also have to recognize, but that's nothing new. In many cases, good LINQ providers match a whole lot of patterns they can translate.

Not supporting Slice or ElementAt in Queryable is also problematic though, because the user will still see these show up because IQueryable<T> derives from IEnumerable<T>, so now there's a silent non-obvious place where one switches from Queryable to Enumerable inside a query expression, which can lead to unexpected execution where a whole lot of query logic ends up being evaluated locally after downloading massive result sets from some database. E.g.

from x in xs.Slice(1..1000)
where f(x)
...

would have where bind to Enumerable.Where because Slice silently took an IQueryable<T> and returned an IEnumerable<T>.

Also, rejecting Queryable APIs is a mood point because one would still see the equivalent Enumerable APIs show up in expression tree contexts, e.g.

from x in xs
group x by g(x) into g
select g.ElementAt(Index.FromEnd(1))

where g is an inner sequence and the select clause's body binds to Enumerable.ElementAt(Index). The result is a MethodCallExpression whose MethodInfo points at a new never seen before ElementAt overload. Query providers already typically detect methods on both Queryable and Enumerable exactly because IEnumerable<T> inner sequences can show up in expression trees.

One final note is that some libraries leverage AsQueryable on enumerable sequences to get an expression tree representation of a query, detect patterns they can optimize by rewriting the tree, and then just evaluate the resulting expression (more efficiently due to the prior rewrites) by enumerating. Lacking these methods on Queryable breaks such cases; throwing in an AsQueryable would result in subsetting the set of methods that's available, effectively creating a dialect.

My 5 cents is to keep Queryable in sync with Enumerbale, live with the C# language restrictions for the time being, and avoid creating a dialect. Silent binding to Enumerable can be harmful (at least having a query provider blow up on an unknown overload does not exhibit silent fallback to local evaluation) and Enumerable methods sneak into query expression trees anyway (so a query provider could still blow up when these methods occur in different positions in the query expression).

**Comment by bartdesmet** [comment]
One note on the EF Core code pointed out. It'd be way more robust to use some InfoOf trick to get these methods:

static MethodInfo InfoOf<T, R>(Expression<Func<T, R>> f) => ((MethodCallExpression)f.Body).Method;

And then use things like

ElementAt = InfoOf((IQueryable<object> source, int index) => source.ElementAt(index)).GetGenericMethodDefinition();

Rather than a bunch of LINQ over reflection APIs without 100% specific filters and shortcuts with Single that make API assumptions. The InfoOf trick effectively gets the token for the method at compile time.

I believe I did put some code in the System.Linq.Expressions lambda compiler that has InfoOf-based tables of well known methods as well (and if memory serves well, it's even just using Func<T, R> rather than expression trees to achieve the result).

**Comment by eiriktsarpalis** [comment]
> assuming xs is an IQueryable, but only if it doesn't occur in an expression tree context, because the C# compiler does not support using ^ or .. in an expression tree (nor do the libraries support FromEnd or Range nodes in expression trees).

That's surprising to me, since I assumed that e.g. the range `1 .. ^2` is simply sugar for the expression `new Range(new Index(1), new Index(2, fromEnd: false))`. Why would we need to extend expression trees here?

**Comment by Dixin** [comment]
## Enumerable vs Queryable vs PLINQ

Thank you all for your time and insights. The reason I include `Queryable` in the proposal is, in my observation, when a LINQ operator is added to `Enumerable`, it is usually added to Queryable, but not added to PLINQ (`ParallelEnumerable`). 

@terrajobst 
> Should we evolve PLinq as well?

In recent years,
- `Append`, `Prepend`, `TakeLast`, `SkipLast`, and a `Zip` overload are added to `Enumerable`. 
- They are all added to `Queryable`. 
- None of them are added to `ParallelEnumerable`. 

I tried to have the proposal follow the existing pattern. But it would be nice if we evolve PLINQ as well.

## Queryable.ElementAt(Index)

@bartdesmet
> My 5 cents is to keep Queryable in sync with Enumerable

I wish they are in sync too. But if `Queryable.ElementAt(Index)` causes problem for LINQ providers, can we only add `ElementAt(Index)` overload to `Enumerable`?

`Min` is such an example. `Min` has 22 overloads on `Enumerable`, but only 2 overloads on `Queryable`. So is `Max`.

@terrajobst @bartdesmet @jogibear9988 @eiriktsarpalis @GrabYourPitchforks Just like `Min` and `Max`, is it an option to have `ElementAt(Index)` and `ElementAt(int)` for `Enumerable`, and keep `Queryable` only with `ElementAt(int)`?

**Comment by bartdesmet** [comment]
@eiriktsarpalis, you've hit on one of the more tricky parts of expression trees. While lowering a from-end or range expression to `Index` and `Range`-based APIs may be possible in some cases, it's neither desirable nor always possible.

Elaborating on the latter case, consider an expression `s[1..^1]` where `s` is an expression of type `string` (or an array for that matter). The libraries do not define indexers with `Range` or `Index` for these types; instead, the language defines these in terms of existing or auxiliary APIs targeting `string` and arrays. For example, `s[^1]` binds to `string[int]` after computing the index. However, for a from-end index, this involves knowing the length of the string, and we only want to evaluate the string once. So lowering the following:

```csharp
Console.ReadLine()[^1]
```

cannot target an `Index`-based overload of an indexer (because such an overload does not exist), and requires the use of a "comma" to evaluate the left-hand side only once, akin to the following pseudo-C#:

```csharp
{ string s; s = Console.ReadLine(); s[s.Length - 1] }
```

While expression trees do have `Block` expressions that can be used to represent this, we know are starting to leak into the territory of statement trees and `Assign` nodes as well. Translation of `s[1..^1]` to a `Substring` call hits similar issues, as does doing the same for arrays (with the added complexity that calls to `RuntimeHelpers.GetSubArray<T>` will also end up leaking into expression trees).

This brings me to the first point stating that such lowering is not desirable either. The original intent of the user, namely specifying a range or an index "from the end", gets lost in translation, making the job of libraries that inspect such intent (e.g. for the purpose of runtime translation to another language, as is the case for query providers) harder. An example of where lowering snuck into expression trees after the original C# 3.0 wave is string interpolation:

```csharp
$"f({x}) = {f(x)}"
```

is turned into a call to `string.Format("f({0}) = {1}", x, f(x))`. No harm done? While it's true that some query providers know of `string.Format`, I've personally cursed at this one when I was targeting a language that had interpolation support itself, leaving me with the task to decompile the format string (so writing a little parser to find the holes with possible format specifiers and alignment specifiers). I'd have been perfectly happy telling users not to use `string.Format` directly (and hence throwing an exception when encountering such a node) and stating that one has to use interpolation instead, if that language construct were captured as an `InterpolatedCSharpExpression`. This is exactly what I've implemented in my [expression trees proposal over here](https://github.com/bartdesmet/ExpressionFutures/blob/master/CSharpExpressions/Microsoft.CSharp.Expressions/Microsoft/CSharp/Expressions/InterpolatedStringCSharpExpression.cs).

At this point, for better or for worse, it's become a routine to reject all new language constructs in expression trees. For example, see https://github.com/dotnet/roslyn/pull/42011. The master issue capturing the lack of evolution of expression trees is over here: https://github.com/dotnet/csharplang/issues/2545. Feel free to upvote or comment in there.

**Comment by jogibear9988** [comment]
@bartdesmet have you also seen https://github.com/dadhi/FastExpressionCompiler
he also added (planed) some expressions to it compiler. So maybe you two could work together when adding new ones to keep the sytax compatible.

**Comment by bartdesmet** [comment]
@jogibear9988, yes, the two are mostly complimentary, with one focusing on the back-end, and another focusing on the front-end. I also do have a bunch more stuff coming up in a few weeks as part of an OSS effort, but I can't elaborate on that just yet.

**Comment by bartdesmet** [comment]
@Dixin, query providers typically do have to understand `Enumerable.*` methods anyway, so they're equally in trouble if a method is just added there. For example:

```csharp
private void F(IQueryable<int> xs)
{
    var res = from x in xs
              group x by x % 2 into g
              // This binds to Enumerable.ElementAt(int).
              select g.ElementAt(1);
}
```

(And similarly if you'd use `Count`, `Max`, `Min`, etc.)

As such, I'd just follow suit with the `Append`, `Prepend`, etc. prior art and add these to `Queryable` as well.

**Comment by eiriktsarpalis** [comment]
@bartdesmet there is a relatively awkward workaround:
```csharp
xs.Slice(1..2).SelectMany(x => ys.Slice(new Range(3,4)))
```
By the sound of it I think we should just go ahead with the Queryable overload and hope that future compiler releases add support for quoted range expressions.

**Comment by bartdesmet** [comment]
@eiriktsarpalis exactly, that's where the expression tree wheels come off today. Sometimes you can manually "lower" the expression, but sometimes you can't because you need temporary assignments, statements, and whatnot (as shown earlier).

**Comment by eiriktsarpalis** [comment]
I'm increasingly warming up to exposing the `Range` method as a `Take` overload, it really simplifies applications that would typically require chaining `Skip` and `Take` calls and the confusion that ensues with order of chaining and indices:

* `source.Take(..3)` instead of `source.Take(3)`
* `source.Take(3..)` instead of `source.Skip(3)`
* `source.Take(2..7)` instead of `source.Skip(2).Take(5)`
* `source.Take(^3..)` instead of `source.TakeLast(3)`
* `source.Take(..^3)` instead of `source.SkipLast(3)`
* `source.Take(^7..^3)` instead of `source.TakeLast(7).SkipLast(3)`. This is also more efficient since elements only need to be queued once instead of twice.

Using the name `Take` also removes any ambiguity in terms of whether out-of-bounds ranges should have `Slice` semantics or `Take` semantics. Clearly, it should be the latter. The semantics of the [implementation as provided by @Dixin](https://github.com/Dixin/System.Linq.IndexRange/blob/c66ba6090cd3bc1dfee8efe1cf359ffadaf617f7/System.Linq.IndexRange/Slice.cs) are just right in that regard.

**Comment by jogibear9988** [comment]
Take(..3)

do slicees start with 0 or 1?

**Comment by eiriktsarpalis** [comment]
> do slicees start with 0 or 1?

They're ranges of indices so they start with 0.

**Comment by jogibear9988** [comment]
so ..3 is not the ssme as take(3) or am i wrong?

**Comment by Dixin** [comment]
> Take(..3)
> 
> do slicees start with 0 or 1?

@jogibear9988 `..3` is compiled to `Range.EndAt(3)`, which returns `new Range(start: new Index(0), end: new Index(3))`

> so ..3 is not the ssme as take(3) or am i wrong?

`Take(..3)` and `Take(3)` work the same.

**Comment by eiriktsarpalis** [comment]
Yeah I should have mentioned that the upper bound is always exclusive (in math terms you'd write it `[0, 3)`).

**Comment by eiriktsarpalis** [comment]
@Dixin would you be able to update the OP to incorporate API review feedback? TL;DR

* We're renaming `Slice` to `Take`.
* We use LINQ behaviour for out-of-bounds ranges.
* We'll be implementing the Queryable equivalents but not the PLINQ ones.

cc @terrajobst

**Comment by Dixin** [comment]
@eiriktsarpalis Thank you for your time. Appended the API review feedback to proposal, also updated proposal contents inline.

**Comment by terrajobst** [comment]
* Looks good as proposed.
* We're not providing an implementation for PLinq
* Query providers can choose implement the method, if no there will be some default behavior.

```C#
namespace System.Linq
{
    public static partial class Enumerable
    {
        public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, Index index);
        public static TSource ElementAtOrDefault<TSource>(this IEnumerable<TSource> source, Index index);
        public static IEnumerable<TSource> Take<TSource>(this IEnumerable<TSource> source, Range range);
    }
    public static partial class Queryable
    {
        public static TSource ElementAt<TSource>(this IQueryable<TSource> source, Index index);
        public static TSource ElementAtOrDefault<TSource>(this IQueryable<TSource> source, Index index);
        public static IQueryable<TSource> Take<TSource>(this IQueryable<TSource> source, Range range);
    }
}
```

**Comment by jogibear9988** [comment]
> EF Core reflection needs to be fixed when Queryable ElementAt is changed: https://github.com/dotnet/efcore/blob/5d537965c5a547c3d6e42b171d7f8a9415554093/src/EFCore/Query/QueryableMethods.cs#L477

With this comment I don't mean EF Core needs to be fixed to support the new one, it needs to be fixed so the original ElementAt will still work. So it means you cannot use current EFCore with Net6 cause the Reflection will fail

**Comment by Dixin** [comment]
@terrajobst @eiriktsarpalis  Thank you for the update. I implemented these APIs. Preview: https://github.com/Dixin/runtime/commit/4f846fedc602682add0f62aa8695cb73a86d76e7
- Enumerable
  - Take: https://github.com/Dixin/runtime/blob/master/src/libraries/System.Linq/src/System/Linq/Take.cs
  - ElementAt: https://github.com/Dixin/runtime/blob/master/src/libraries/System.Linq/src/System/Linq/ElementAt.cs
- Queryable: https://github.com/Dixin/runtime/blob/master/src/libraries/System.Linq.Queryable/src/System/Linq/Queryable.cs

I am working on unit tests, etc.

**Comment by eiriktsarpalis** [comment]
@Dixin feel free to submit a PR if you're working on the implementation.

**Comment by Dixin** [comment]
@eiriktsarpalis I will submit PR once I am done with the unit tests. I am paralyzed, please expect slower speed.

**Comment by eiriktsarpalis** [comment]
That's absolutely fine.

**Comment by Dixin** [comment]
I am finalizing the code https://github.com/Dixin/runtime/commits/linq-index-range. However, I cannot find a place to add the XML comments `/// <summary>...` for the new APIs. I searched the entire repo, cannot find a proper file to add the XML documentation for `Enumerable` and `Queryable`. 

@eiriktsarpalis Could you please help?

**Comment by stephentoub** [comment]
> I cannot find a place to add the XML comments

Just add them as /// comments directly on the public APIs you're adding in the src.

**Comment by Dixin** [comment]
@stephentoub Thank you for your reply. The file in `src` contains no `/// comments` for existing APIs at all:
- https://github.com/dotnet/runtime/blob/master/src/libraries/System.Linq/src/System/Linq/ElementAt.cs
- https://github.com/dotnet/runtime/blob/master/src/libraries/System.Linq/src/System/Linq/Take.cs

Are you sure about adding `/// comments` in these files?

The file in `ref` contains no `/// comments` for existing APIs either:
- https://github.com/dotnet/runtime/blob/master/src/libraries/System.Linq/ref/System.Linq.cs

**Comment by stephentoub** [comment]
> Are you sure about adding /// comments in these files?

Yes.  We're documenting all new APIs with /// comments to seed the documentation, and we're on a path to moving all documentation into the src files as well so that they may become the "source of truth" for the docs.

**Comment by eiriktsarpalis** [comment]
FWIW our current source of truth [can be found here](https://github.com/dotnet/dotnet-api-docs/tree/master/xml/System.Linq) if you're looking for a template.
